backend/local/local.go

```diff
@@ -494,92 +496,145 @@ func (f *Fs) List(ctx context.Context, dir string) (entries fs.DirEntries, err e
+
+		g, gCtx := errgroup.WithContext(ctx)
+		g.SetLimit(fs.GetConfig(ctx).Checkers)
+		for i, fi := range fis {
+			if fi == nil {
+				continue
 			}
-			if fi.IsDir() {
-				// Ignore directories which are symlinks.  These are junction points under windows which
-				// are kind of a souped up symlink. Unix doesn't have directories which are symlinks.
-				if (mode&os.ModeSymlink) == 0 && f.dev == readDevice(fi, f.opt.OneFileSystem) {
-					d := fs.NewDir(newRemote, fi.ModTime())
-					entries = append(entries, d)
+			i, fi := i, fi // https://golang.org/doc/faq#closures_and_goroutines
+			g.Go(func() error {
+				// No point in continuing if context has been cancelled
+				if gCtx.Err() != nil {
+					return nil
 				}
-			} else {
-				// Check whether this link should be translated
-				if f.opt.TranslateSymlinks && fi.Mode()&os.ModeSymlink != 0 {
-					newRemote += linkSuffix
-				}
-				// Don't include non directory if not included
-				// we leave directory filtering to the layer above
-				if useFilter && !filter.IncludeRemote(newRemote) {
-					continue
+
+				var err error
+				name := fi.Name()
+				mode := fi.Mode()
+				newRemote := f.cleanRemote(dir, name)
+
+				// Follow symlinks if required
+				if f.opt.FollowSymlinks && (mode&os.ModeSymlink) != 0 {
+					localPath := filepath.Join(fsDirPath, name)
+					fi, err = os.Stat(localPath)
+					if err != nil {
+						// Quietly skip errors on excluded files and directories
+						if useFilter && !filter.IncludeRemote(newRemote) {
+							return nil
+						}
+						if os.IsNotExist(err) || isCircularSymlinkError(err) {
+							// Skip bad symlinks and circular symlinks
+							err = fserrors.NoRetryError(fmt.Errorf("symlink: %w", err))
+							fs.Errorf(newRemote, "Listing error: %v", err)
+							_ = accounting.Stats(gCtx).Error(err)
+							return nil
+						}
+						return err
+					}
+					mode = fi.Mode()
 				}
-				fso, err := f.newObjectWithInfo(newRemote, fi)
-				if err != nil {
-					return nil, err
+
+				// No point in continuing if context has been cancelled
+				if gCtx.Err() != nil {
+					return nil
 				}
-				if fso.Storable() {
-					entries = append(entries, fso)
+
+				if fi.IsDir() {
+					// Ignore directories which are symlinks.  These are junction points under windows which
+					// are kind of a souped up symlink. Unix doesn't have directories which are symlinks.
+					if (mode&os.ModeSymlink) == 0 && f.dev == readDevice(fi, f.opt.OneFileSystem) {
+						d := fs.NewDir(newRemote, fi.ModTime())
+						loopEntries[i] = d
+						return nil
+					}
+				} else {
+					// Check whether this link should be translated
+					if f.opt.TranslateSymlinks && fi.Mode()&os.ModeSymlink != 0 {
+						newRemote += linkSuffix
+					}
+					// Don't include non directory if not included
+					// we leave directory filtering to the layer above
+					if useFilter && !filter.IncludeRemote(newRemote) {
+						return nil
+					}
+					fso, err := f.newObjectWithInfo(newRemote, fi)
```

Comment:
There also seems to be lock contention here for the lock on `fs.objectMetaMu`, but I haven't been able to understand why it's a global fs-level lock rather than an object-specific lock. All it's uses seem to be for object-specific metadata.

Reply:
Its a global because I didn't want to make the Object struct any bigger. However in needs to be a local then I suggest you add another commit to this PR making it local to Object. We can then see if that helps.